allenact/algorithms/onpolicy_sync/engine.py

```diff
@@ -1412,34 +1452,71 @@ def run_pipeline(self, rollouts: RolloutStorage):
                 ndone = int(self.num_workers_done.get("done"))
                 assert (
                     ndone == self.num_workers
-                ), "# workers done {} != # workers {}".format(ndone, self.num_workers)
+                ), f"# workers done {ndone} != # workers {self.num_workers}"
 
                 # get the actual step_count
                 self.step_count = (
                     int(self.num_workers_steps.get("steps")) + self.former_steps
                 )
 
-            rollouts.compute_returns(
+            before_update_info = dict(
                 next_value=actor_critic_output.values.detach(),
-                use_gae=self.training_pipeline.current_stage.use_gae,
-                gamma=self.training_pipeline.current_stage.gamma,
-                tau=self.training_pipeline.current_stage.gae_lambda,
+                use_gae=cur_stage_training_settings.use_gae,
+                gamma=cur_stage_training_settings.gamma,
+                tau=cur_stage_training_settings.gae_lambda,
+                adv_stats_callback=self.advantage_stats,
             )
 
-            self.update(rollouts=rollouts)  # here we synchronize
-            self.training_pipeline.rollout_count += 1
+            # Prepare storage for iteration during updates
+            for storage in self.training_pipeline.current_stage_storage.values():
+                storage.before_updates(**before_update_info)
+
+            for sc in self.training_pipeline.current_stage.stage_components:
+                component_storage = uuid_to_storage[sc.storage_uuid]
+
+                # before_update = time.time()
+
+                self.update(
+                    stage=self.training_pipeline.current_stage,
+                    stage_component=sc,
+                    storage=component_storage,
+                )
 
-            rollouts.after_update()
+                # after_update = time.time()
```

Comment:
Yah, I wasn't sure if I should delete it or no. It might be nice to have a "benchmarking" context that will only run certain code if we set a particular flag.